support-models/src/main/scala/com/gu/support/workers/GiftRecipient.scala

```diff
@@ -54,6 +55,53 @@ object GiftRecipient {
     ))
 }
 
+case class GiftCode private(value: String) extends AnyVal
+
+object GiftCode {
+
+  def apply(value: String): Option[GiftCode] =
+    Some(value)
+      .filter(_.matches(raw"""gd(03|06|12)-[a-km-z02-9]{8}"""))
+      .map(new GiftCode(_))
+
+  implicit val e1 = Encoder.encodeString.contramap[GiftCode](_.value)
+  implicit val d1 = Decoder.decodeString.emap(GiftCode(_).toRight("invalid gift code"))
+
+}
+
+sealed trait GiftRecipientAndMaybeCode {
```

Comment:
I am going to rename this, something like `SendThankYouEmailGiftPurchase` is a more meaningful name, so far it's called by the "what" rather than the "why"

Reply:
```
GiftPurchase
+- WeeklyGiftPurchase
+- DigitalSubGiftPurchase
```